### PR TITLE
non operator able to target and ssh node through bastion host

### LINKED ISSUE #233: As non garden admin I am not able to target shoots
**Describe the bug**
I am not an admin in this garden but wanted to use gardenctl to manage the shoots in my project. I can target my project, but not my shoots.

**To Reproduce**
1. Which target was set 'gardenctl get target'

```
- kind: project
  name: project-xyz
```
2. Which command was entered [e.g. 'gardenctl show vpn-seed']
`gardenctl target shoot shoot-xyz`
3. What was the output of the command 
```
shoots.core.gardener.cloud "shoot-xyz" is forbidden: User "system:serviceaccount:garden-xyz:xyz-sa" cannot list resource "shoots" in API group "core.gardener.cloud" at the cluster scope
```

**Expected behavior**
I am able to target the shoots I have admin permissions to and use e.g. gardenctl ssh for it.

**Gardenctl Version (please complete the following information):**
```
gardenctl:
		version     : v0.16.0
		build date  : 2020-01-28
		go version  : go1.13.6
		go compiler : gc
		platform    : darwin/amd64
```

### LINKED ISSUE #253: As non garden admin I am not able to get logs or ssh into a node
**Describe the bug**
I am not an admin in this garden but wanted to use gardenctl to manage the shoots in my project. With the fix in https://github.com/gardener/gardenctl/pull/249 I am able to target my shoot, but I can't do anything with it like ssh into a node or retrieving logs. 

**To Reproduce**
1. Which target was set 'gardenctl get target'
```
- kind: shoot
  name: cds-b-sus
```
2. Which command was entered [e.g. 'gardenctl show vpn-seed']
`gardentctl logs vpn-shoot` and `gardentctl ssh ip-a-b-c-d.ec2.internal`

3. What was the output of the command 

```
logs
shoots.core.gardener.cloud is forbidden: User "system:serviceaccount:garden-x:x-sa" cannot list resource "shoots" in API group "core.gardener.cloud" at the cluster scope

ssh
secrets "seed-aws-us1" is forbidden: User "system:serviceaccount:garden-x:x-sa" cannot get resource "secrets" in API group "" in the namespace "garden"
```
**Expected behavior**
SSH into shoot nodes or retrieving logs of vpn-shoot works.

**Gardenctl Version (please complete the following information):**
```
gardenctl:
		version     : v0.20.0
		build date  : 2020-08-10
		go version  : go1.14.6
		go compiler : gc
		platform    : darwin/amd64
```

### PR DESCRIPTION
**Pull request by tedteng**
**What this PR does / why we need it**:
allow non-operator user use project SA account in garden able to manage the shoots in their project via `gardenctl target` and `gardenctl ssh` to list nodes ssh shoot nodes through bastion host. 

**Which issue(s) this PR fixes**:
Fixes https://github.com/gardener/gardenctl/issues/253
Fixes https://github.com/gardener/gardenctl/issues/233

**Special notes for your reviewer**:
as a request from the end-user currently the `ssh`  only allow access aws provider.

**Release note**:

```improvement user
Allow `user` role use Project SA account in garden able to manage the shoots from their own project. `gardenctl ssh` shoot node as well.  `aws`,`gcp`,`azure` supported
```

### CODE REVIEW

**Comment by tedteng** [comment]
Testing 
```
gg target -g ttt -p i333xxx -t zegfzwjhzh
Garden:
KUBECONFIG=/Users/i333xxx/gardener_kubeconfig/ttt/kubeconfig

Warning:
You are non-operator! usage are Limited

Shoot:
KUBECONFIG=/Users/i333878/.garden/cache/ttt/projects/i333878/zegfzwjhzh/kubeconfig.yaml
```

`gg ssh` list the node as expected. 

`gg ssh  ip-xxxxx.eu-central-1.compute.internal`

**Review by petersutter — CHANGES_REQUESTED** [review]

File: pkg/cmd/miscellaneous.go
Comment (tedteng) [reply pkg/cmd/miscellaneous.go]: fixed, remove this method

**Comment by tedteng** [comment]
> Please do not amend your commits and force push. Instead just commit every change you do so that I can better keep track what is new and what you have changed, thanks.
> 
> Once we are done with the review you can then squash it finally.

didn't realize that, Thanks for the remind

File: pkg/cmd/target.go
Comment (tedteng) [reply pkg/cmd/target.go:579]: open issue https://github.com/gardener/gardenctl/issues/269 under discussion

File: pkg/cmd/operate.go
Comment (petersutter) [reply pkg/cmd/operate.go:87]: As now already mentioned multiple times: Do not list shoots across all namespaces (cluster scope). You do not have the permission to do that as you can see also in your screenshot.

I have also already suggested what should be done instead, see snipped here: https://github.com/gardener/gardenctl/pull/263/files#r474740207

File: pkg/cmd/target.go
Comment (petersutter) [inline pkg/cmd/target.go:570]:
```suggestion
	// temporary solution , will clean up code in ticket move get seed out of targetShoot method #269
```
added reference to the issue

File: pkg/cmd/target.go
Comment (petersutter) [inline pkg/cmd/target.go:686]:
```suggestion
	// temporary solution , will clean up code in ticket move get seed out of targetShoot method #269
```
added reference to the issue

File: pkg/cmd/operate.go
Comment (tedteng) [reply pkg/cmd/operate.go:87]: sure I haven't completed this part yet.  will fix it

File: pkg/cmd/ssh_aws.go
Comment (petersutter) [inline pkg/cmd/ssh_aws.go:104]:
```go
a.ShootName = "shoot--" + target.Stack()[1].Name + "--" + target.Stack()[2].Name
```
Don't do that. Use `shoot.Status.TechnicalID` as I wrote before

File: pkg/cmd/ssh_aws.go
Comment (petersutter) [inline pkg/cmd/ssh_aws.go:142]: same here, use `shoot.Status.TechnicalID`

**Review by petersutter — CHANGES_REQUESTED** [review]

File: pkg/cmd/miscellaneous.go
Comment (petersutter) [inline pkg/cmd/miscellaneous.go:379]:
```go
return string(shoot.Status.TechnicalID)
```
TechnicalID is already a string, no need to stringify it

Comment (tedteng) [reply pkg/cmd/miscellaneous.go]: done

File: pkg/cmd/miscellaneous.go
Comment (petersutter) [inline pkg/cmd/miscellaneous.go:369]: what about the `getShootClusterName` function? It should be removed and this function should be used instead

Comment (tedteng) [reply pkg/cmd/miscellaneous.go:344]: done

File: pkg/cmd/operate.go
Comment (tedteng) [reply pkg/cmd/operate.go]: done

File: pkg/cmd/miscellaneous.go
Comment (tedteng) [reply pkg/cmd/miscellaneous.go]: done

File: pkg/cmd/download.go
Comment (tedteng) [reply pkg/cmd/download.go]: fixed

**Comment by tedteng** [comment]
@petersutter all fixed now.  let me know any concerns.

File: pkg/cmd/miscellaneous.go
Comment (petersutter) [reply pkg/cmd/miscellaneous.go:330]: please commit my suggestion, do not introduce for new code this bash hacking

File: pkg/cmd/ssh.go
Comment (petersutter) [reply pkg/cmd/ssh.go:150]: at least call kubectl directly, no bash please for new code

File: pkg/cmd/ssh.go
Comment (petersutter) [reply pkg/cmd/ssh.go]: do not call bash

File: pkg/cmd/miscellaneous.go
Comment (petersutter) [reply pkg/cmd/miscellaneous.go:320]: do not call bash

File: pkg/cmd/ssh.go
Comment (tedteng) [reply pkg/cmd/ssh.go]: fixed

File: pkg/cmd/ssh.go
Comment (tedteng) [reply pkg/cmd/ssh.go]: fixed

File: pkg/cmd/miscellaneous.go
Comment (tedteng) [reply pkg/cmd/miscellaneous.go]: done

File: pkg/cmd/miscellaneous.go
Comment (tedteng) [reply pkg/cmd/miscellaneous.go]: done

**Review by petersutter — APPROVED** [review]
/lgtm thanks

the code looks good to me however I haven't tested it

**Comment by tedteng** [comment]
> /lgtm thanks
> 
> the code looks good to me however I haven't tested it

Thanks your help, I will check it after the merge, meanwhile fix the conflict as well

**Comment by petersutter** [comment]
@tedteng I just checked the release notes that you have added. Shouldn't it also mention that `gardenctl ssh` currently only works for `aws`?

**Comment by tedteng** [comment]
> @tedteng I just checked the release notes that you have added. Shouldn't it also mention that `gardenctl ssh` currently only works for `aws`?

yes updated, will working on others later

**Comment by tedteng** [comment]
try to squash-commits , hopeful everything is good.  let me know if there any wrong.

**Comment by tedteng** [comment]
Hi let me know is the commit square ok. @gardenctl-maintainers

**Comment by neo-liang-sap** [comment]
/lgtm 
looks good to me but i haven't test it

**Comment by neo-liang-sap** [comment]
@tedteng could you please squash commits into one?

**Comment by tedteng** [comment]
you mean all the commits for just two commits from mine?

**Comment by neo-liang-sap** [comment]
> you mean all the commits for just two commits from mine?

why other people's commit exist in your PR? ideally before merge one PR should contains only one commit from your self.
maybe you can try rebase with master branch but i'm not sure whether it helps to tidy your commits.
also i saw one `empty` commit exist in your PR, https://github.com/gardener/gardenctl/pull/263/commits/51b652710011a1e9b2695da9f239239845ac3ebf not sure how it comes but

**Comment by tedteng** [comment]
I have no idea either that why I was asking check last week.  let me just close and open new PR, I think it will be easier for all. 

> > you mean all the commits for just two commits from mine?
> 
> why other people's commit exist in your PR? ideally before merge one PR should contains only one commit from your self.
> maybe you can try rebase with master branch but i'm not sure whether it helps to tidy your commits.
> also i saw one `empty` commit exist in your PR, [51b6527](https://github.com/gardener/gardenctl/commit/51b652710011a1e9b2695da9f239239845ac3ebf) not sure how it comes but it shouldn't be there

**Review by neo-liang-sap — CHANGES_REQUESTED** [review]
1) please resolve conflict with master branch 
2) these 4 commits are already in master branch, should not exist in this PR
image

**Review by neo-liang-sap — APPROVED** [review]
/lgtm
however i haven't tested, confirmed with @tedteng he's fully confidence with it

**Pull request merged**
